File: leetcode/fibonacci-number.py
dp = {}

class Solution:
    def fib(self, n: int) -> int:
        if dp.get(n):
            return dp[n]
        if n == 0:
            return 0
        if n == 1:
            return 1

        a = Solution()
        c = a.fib(n - 1) + a.fib(n - 2)
        dp[n] = c
        return c

File: leetcode/fibonacci-number.cpp
class Solution {
public:
    int fib(int n) {
        map<int, int> m;
        function<int (int)> f = [&](int x) {
            if (x == 0) return 0;
            if (x == 1) return 1;
            if (m.count(x)) return m[x];
            return m[x] = f(x - 1) + f(x - 2);
        };
        return f(n);
    }
};